cmake/Dependencies.cmake

```diff
@@ -274,44 +274,59 @@ if(INTERN_BUILD_MOBILE AND INTERN_USE_EIGEN_BLAS)
 endif()
 
 # ---[ pthreadpool
-# QNNPACK and NNPACK both depend on pthreadpool, but when building with libtorch
-# they should use the pthreadpool implementation under caffe2/utils/threadpool
-# instead of the default implementation. To avoid confusion, add pthreadpool
-# subdirectory explicitly with EXCLUDE_FROM_ALL property prior to QNNPACK/NNPACK
-# does so, which will prevent it from installing the default pthreadpool library.
-if(INTERN_BUILD_MOBILE AND NOT BUILD_CAFFE2_MOBILE AND (USE_QNNPACK OR USE_NNPACK OR USE_XNNPACK))
-  if(NOT DEFINED PTHREADPOOL_SOURCE_DIR)
-    set(CAFFE2_THIRD_PARTY_ROOT "${PROJECT_SOURCE_DIR}/third_party")
-    set(PTHREADPOOL_SOURCE_DIR "${CAFFE2_THIRD_PARTY_ROOT}/pthreadpool" CACHE STRING "pthreadpool source directory")
-  endif()
+# pthreadpool is a C library with no concept of namespaces, which means two
+# copies of the library cannot exist in the same binary or symbol collision
+# will occur violating ODR.  The conundrum is that Caffe2 provides its own
+# custom implementation of pthreadpool under the exact same interface the open
+# source version operates under; an implementation that is only used in Caffe2
+# operators and Caffe2's usage of NNPACK and QNNPACK. On the other hand, we
+# are interested in using the open source implementation on PyTorch which seems
+# to have been updated to exhibit superior performance.  Here we try to solve
+# this conundrum by prefering to pick the internal Caffe2 implementation if
+# BUILD_CAFFE2_OPS is on.  This will force both PyTorch and Caffe2 to use the
+# Caffe2 pthreadpool version (remember we cannot have two copies in one binary)
+# and in doing so we achieve our goal of keeping Caffe2's behavior intact if
+# at the cost of some performance loss to PyTorch.  Otherwise, if BUILD_CAFFE2_OPS
+# is not requested, we take the liberty of picking the open source version to
+# improve PyTorch's performance.  All of this admittedly error prone logic
+# can be removed and simplified if and when we decide to migrate Caffe2 to the
+# open source pthreadpool as well.
+
+if(NOT USE_SYSTEM_PTHREADPOOL)
```

Comment:
A contributor recently added USE_SYSTEM_PTHREADPOOL to mean the one installed on the Linux distribution.  So we now have the choice between C2's implementation, the system implementation, and third party implementation!  This complicates everything.  I am doing my best to handle it here.

Reply:
@cdluminate Please review.